app/src/main/java/com/github/swent/echo/compose/authentication/AuthenticationForm.kt

```diff
@@ -28,47 +31,133 @@ import com.github.swent.echo.R
-        OutlinedTextField(
+        AuthenticationTextField(
             modifier = Modifier.fillMaxWidth().testTag("email-field"),
-            label = { Text(stringResource(R.string.authentication_form_email_label)) },
+            label = stringResource(R.string.authentication_form_email_label),
             value = email,
-            onValueChange = { email = it },
-            singleLine = true
+            onValueChange = {
+                emailError = ""
+                email = it
+            },
+            errorText = emailError,
         )
-        Spacer(modifier = Modifier.padding(16.dp))
-        OutlinedTextField(
+        Spacer(modifier = Modifier.padding(8.dp))
+        AuthenticationTextField(
             modifier = Modifier.fillMaxWidth().testTag("password-field"),
-            label = { Text(stringResource(R.string.authentication_form_password_label)) },
+            label = stringResource(R.string.authentication_form_password_label),
             value = password,
-            onValueChange = { password = it },
-            visualTransformation = PasswordVisualTransformation(),
-            keyboardOptions =
-                KeyboardOptions(
-                    keyboardType = KeyboardType.Password,
-                ),
-            singleLine = true
+            onValueChange = {
+                passwordError = ""
+                password = it
+            },
+            errorText = passwordError,
+            isPasswordField = true,
         )
-        Spacer(modifier = Modifier.padding(16.dp))
+        if (confirmPassword) {
+            Spacer(modifier = Modifier.padding(8.dp))
+            AuthenticationTextField(
+                modifier = Modifier.fillMaxWidth().testTag("confirm-password-field"),
+                label = stringResource(R.string.authentication_form_confirm_password_label),
+                value = confirm,
+                onValueChange = {
+                    confirmError = ""
+                    confirm = it
+                },
+                errorText = confirmError,
+                isPasswordField = true,
+            )
+        }
+        Spacer(modifier = Modifier.padding(8.dp))
         ElevatedButton(
             enabled = isOnline,
             modifier = Modifier.fillMaxWidth().testTag("action-button"),
-            onClick = { onAuthenticate(email, password) },
+            onClick = {
+                if (validate) {
+                    if (email.isEmpty()) {
+                        emailError = "Email is required"
+                    }
+
+                    if (password.length < 8) {
+                        passwordError = "Password must be at least 8 characters"
+                    }
+
+                    if (confirmPassword && password != confirm && passwordError.isEmpty()) {
+                        confirmError = "Passwords do not match"
+                    }
+
+                    if (emailError.isEmpty() && passwordError.isEmpty() && confirmError.isEmpty()) {
```

Comment:
I feel like having 3 differents strings to handle the errors is a bit repetitive and non-practical if we have to add new errors or modifying the text message in the future. Also using strings to see errors is a bit weird. I think it would be better to create an enum with the different type of errors that contains a message, and then create a unique set of errors that you feed in the above if's. This way, you just have to test if this set is empty in the last condition, and the code is easier to maintain.